crates/rpc-server/src/facade.rs

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2024 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::fmt::Display;
+use std::sync::Arc;
+use std::time::Duration;
+
+use ccp_shared::nox_ccp_api::NoxCCPApi;
+use ccp_shared::proof::CCProof;
+use ccp_shared::proof::ProofIdx;
+use ccp_shared::types::CUAllocation;
+use ccp_shared::types::EpochParameters;
+
+use eyre::Context;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::TryRecvError;
+use tokio::sync::Mutex;
+use tokio::time::timeout;
+
+const WAIT_LOCK_DURATION: Duration = Duration::from_secs(2);
+
+/// An async façade for RPC.
+pub struct OfflineFacade<P> {
+    sender_to_worker: mpsc::Sender<FacadeMessage>,
+    prover: Arc<Mutex<P>>,
+}
+
+impl<P> OfflineFacade<P>
+where
+    P: NoxCCPApi + 'static,
+    <P as NoxCCPApi>::Error: Display,
+{
+    pub fn new(prover: P) -> Self {
+        let prover = Arc::new(Mutex::new(prover));
+
+        let (sender_to_worker, receiver) = mpsc::channel(100);
+
+        let _worker = tokio::task::spawn(facade_loop(prover.clone(), receiver));
```

Comment:
It's usually better to have a dedicated `stop` method, which receives prover by value, drop channel in it, and then wait for a tokio task handle.